pkg/blockbuilder/blockbuilder_test.go

```diff
@@ -287,6 +280,121 @@ func TestBlockBuilder_WithMultipleTenants(t *testing.T) {
+	require.NoError(t, err)
+
+	// We don't want to run the service here. Instead, the test cases below trigger and assert the consumption cycles explicitly.
+	require.NoError(t, bb.starting(ctx))
+	t.Cleanup(func() {
+		require.NoError(t, bb.stopping(nil))
+	})
+
+	const tenantID = "1"
+
+	bucketDir := path.Join(cfg.BlocksStorage.Bucket.Filesystem.Directory, tenantID)
+
+	cycleEndStartup := cycleEndAtStartup(time.Now(), cfg.ConsumeInterval, cfg.ConsumeIntervalBuffer)
+	cycleEnd := cycleEndStartup
+
+	var allSamples []mimirpb.Sample
+
+	// Out of order sample w.r.t. samples in last cycle. But for this cycle,
+	// the TSDB starts fresh. So in terms of the actual block building, it will be
+	// taken as in-order. Depending on if the out-of-order and in-order sample below
+	// crosses the 2h block boundary, we expect either 1 or 2 blocks.
+	{
+		kafkaRecTime := cycleEnd
+		outOfOrderSampleTime := kafkaRecTime.Add(-time.Hour)
+		samples := produceSamples(ctx, t, kafkaClient, kafkaRecTime, tenantID, outOfOrderSampleTime)
+		allSamples = append(allSamples, samples...)
+	}
+
+	// In-order sample w.r.t. last consume cycle.
+	{
+		kafkaRecTime := cycleEnd.Add(cfg.ConsumeInterval / 10)
+		inOrderSampleTime := kafkaRecTime.Add(-time.Minute)
+		samples := produceSamples(ctx, t, kafkaClient, kafkaRecTime, tenantID, inOrderSampleTime)
+		allSamples = append(allSamples, samples...)
+	}
+
+	// Advance the tracking time to the next cycle.
+	cycleEnd = cycleEnd.Add(cfg.ConsumeInterval)
+
+	// Sample is not a part of the next cycle (a future sample) but its kafka record falls in this cycle.
+	// This sample should not go into the first cycle.
+	{
+		kafkaRecTime := cycleEnd.Add(cfg.ConsumeInterval / 10)
+		sampleTime := cycleEnd.Add(cfg.ConsumeInterval)
+		samples := produceSamples(ctx, t, kafkaClient, kafkaRecTime, tenantID, sampleTime)
+		allSamples = append(allSamples, samples...)
+	}
+
+	// In-order sample falls within the next cycle, but the kafka record time does not fall within the next consume cycle.
+	// This sample should not go into the first cycle.
+	{
+		kafkaRecTime := cycleEnd.Add(time.Minute)
+		sampleTime := cycleEndStartup.Add(cfg.ConsumeInterval / 10).Add(time.Minute)
+		samples := produceSamples(ctx, t, kafkaClient, kafkaRecTime, tenantID, sampleTime)
+		allSamples = append(allSamples, samples...)
+	}
+
+	t.Run("consume only out of order and in-order", func(t *testing.T) {
+		require.NoError(t, bb.nextConsumeCycle(ctx, cycleEnd))
+
+		compareQuery(t,
+			mustTSDBOpen(t, bucketDir),
+			allSamples[:len(allSamples)-2], // Don't expect the last two sample.
+			nil,
+			labels.MustNewMatcher(labels.MatchRegexp, "foo", ".*"),
+		)
+	})
+
+	t.Run("future record", func(t *testing.T) {
+		// The sample from above which was in-order but the kafka record was in future
+		// should get consumed in this cycle. The other sample that is still in the future should not be consumed.
+		cycleEnd = cycleEnd.Add(cfg.ConsumeInterval)
```

Comment:
doing this ourselves skips testing the BB logic which does it. Don't we want to test that too? 

Can we let the BB run until it has filled the bucket with some data (i.e. we expect 3 blocks; do something like `Eventually(func() {bucket.countBlocks() == 3})` or until it updates some metrics. Then we can assert on what each block contains with `tsdb.OpenBlock()` instead of `tsdb.Open`

Thoughts?

Reply:
I didn't get the point, sorry. Here, what we do is, effectively, move the wall clock forward. It's not that we "skip" any of the block-builder's logic. We only point it to a portion of the partition, where the cycle's data represent what the test case tests.

We would get the same if we start block-builder and let it run for several hours. It would go over all the test data over the course of multiple cycle hours, and tested the data produced on each cycle.

Reply:
> We would get the same if we started the block-builder and let it run for several hours

yes, but we'd exercise the production code logic. Right now the logic for calculating `cycleEnd` is in both the tests and the prod code.

but your point about waiting for hours also makes sense. Maybe smaller blocks can solve this? like 2-3-second long blocks with a much smaller ConsumeInterval? A small test is better than no test, so i don't want to block on this